django/curator/tests/test_tag_deduplication.py

```diff
@@ -0,0 +1,117 @@
+import random
+import shortuuid
+import string
+
+from django.test import TestCase
+from taggit.models import Tag
+
+from curator.tag_deduplication import TagPreprocessing, TagClustering, TagGazetteering
+from curator.models import CanonicalTag, CanonicalTagMapping
+
+
+class TagPreprocessingTestCase(TestCase):
+    def setUp(self):
+        pass
+
+    # Make sure tag preprocessing
+    def test_split_by_and(self):
+        preprocessed_Tags = TagPreprocessing.preprocess("biology and chemistry")
+        self.assertEqual(
+            ["biology", "chemistry"], preprocessed_Tags, "Failed to split by 'and'"
+        )
+        self.assertEqual(list, type(preprocessed_Tags))
+
+    def test_split_by_semicolon(self):
+        preprocessed_Tags = TagPreprocessing.preprocess("agent-based model; land-use")
+        self.assertEqual(
+            ["agent-based model", "land-use"],
+            preprocessed_Tags,
+            "Failed to split by ';'",
+        )
+        self.assertEqual(list, type(preprocessed_Tags))
+
+    def test_split_by_comma(self):
+        preprocessed_Tags = TagPreprocessing.preprocess("agent-based model, land-use")
+        self.assertEqual(
+            ["agent-based model", "land-use"],
+            preprocessed_Tags,
+            "Failed to split by ','",
+        )
+        self.assertEqual(list, type(preprocessed_Tags))
+
+
+class TestTagClustering(TestCase):
+    def setUp(self):
+        for i in range(300):
+            id = TestTagClustering._id_generator()
+            CanonicalTag(name=id).save()
```

Comment:
this would be more efficient if you create these objects directly and use [bulk_create](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#bulk-create) to persist them to the database

also, `id` is a [builtin Python function](https://docs.python.org/3/library/functions.html) and shadowing these (or any other standard library modules) is generally not recommended - that's why I changed the other one to `_id` but I missed this section of code